internal/cache/file/downloader/parallel_downloads_job.go

```diff
@@ -58,61 +60,158 @@ func (job *Job) downloadRange(ctx context.Context, dstWriter io.Writer, start, e
+	// R.
+	job.mu.Lock()
+	defer job.mu.Unlock()
+
+	finalStart := offsetStart
+	finalEnd := offsetEnd
+
+	if offsetStart != 0 {
+		leftStart, exist := rangeMap[offsetStart]
+		if exist {
+			finalStart = leftStart
+			delete(rangeMap, offsetStart)
+			delete(rangeMap, leftStart)
+		}
+	}
+
+	rightEnd, exist := rangeMap[offsetEnd]
+	if exist {
+		finalEnd = rightEnd
+		delete(rangeMap, offsetEnd)
+		delete(rangeMap, rightEnd)
+	}
+
+	rangeMap[finalStart] = finalEnd
+	rangeMap[finalEnd] = finalStart
+
+	if finalStart == 0 {
+		updateErr := job.updateStatusOffset(finalEnd)
+		if updateErr != nil {
+			return updateErr
+		}
+	}
+	return nil
+}
+
+func (job *Job) downloadOffsets(ctx context.Context, goroutineIndex int64, cacheFile *os.File, rangeMap map[uint64]uint64) func() error {
+	return func() error {
+		// Since we keep a goroutine for each job irrespective of the maxParallelism,
+		// not releasing the default goroutine to the pool.
+		if goroutineIndex > 0 {
+			defer job.maxParallelismSem.Release(1)
+		}
+
+		for {
+			// Read the offset to be downloaded from the channel.
+			objectRange, ok := <-job.offsetChan
+			if !ok {
+				// In case channel is closed return.
+				return nil
+			}
+
+			offsetWriter := io.NewOffsetWriter(cacheFile, int64(objectRange.Start))
+			err := job.downloadRange(ctx, offsetWriter, objectRange.Start, objectRange.End)
+			if err != nil {
+				return err
+			}
+
+			err = job.updateRangeMap(rangeMap, objectRange.Start, objectRange.End)
+			if err != nil {
+				return err
+			}
+		}
+	}
+}
+
 // parallelDownloadObjectToFile does parallel download of the backing GCS object
 // into given file handle using multiple NewReader method of gcs.Bucket running
 // in parallel. This function is canceled if job.cancelCtx is canceled.
 func (job *Job) parallelDownloadObjectToFile(cacheFile *os.File) (err error) {
-	var start, end int64
-	end = int64(job.object.Size)
-	var parallelReadRequestSize = int64(job.fileCacheConfig.DownloadChunkSizeMb) * cacheutil.MiB
-
-	// Each iteration of this for loop downloads job.fileCacheConfig.DownloadChunkSizeMb * job.fileCacheConfig.ParallelDownloadsPerFile
-	// size of range of object from GCS into given file handle and updates the
-	// file info cache.
-	for start < end {
-		downloadErrGroup, downloadErrGroupCtx := errgroup.WithContext(job.cancelCtx)
-
-		for goRoutineIdx := int64(0); (goRoutineIdx < job.fileCacheConfig.ParallelDownloadsPerFile) && (start < end); goRoutineIdx++ {
-			rangeStart := start
-			rangeEnd := min(rangeStart+parallelReadRequestSize, end)
-			currGoRoutineIdx := goRoutineIdx
-
-			// Respect max download parallelism only beyond first go routine.
-			if currGoRoutineIdx > 0 && !job.maxParallelismSem.TryAcquire(1) {
-				break
-			}
+	rangeMap := make(map[uint64]uint64)
+	// Trying to keep the channel size greater than PrallelDownloadsPerFile to ensure
+	// that there is no goroutine waiting for data.
```

Comment:
nit: Trying to keep -> Keeping.
no goroutine waiting for data -> no goroutine waiting for next range to download.

Reply:
1. Trying is already verb. I dont we can use trying to keeping.
2. Updated